[Ben Joseph Murphy] iP

benmurphyy opened the pull request:
# Duke
 - text-based
 - cool
 - **fun**

> “Have fun, even if it’s not the same kind of fun everyone else is having.”
― C.S. Lewis [Source](https://shakejump.com/having-fun-quotes/)

Complete these steps to start your adventure with Duke!
1. Clone this repo
2. Run the main method in Duke class 
3. Have **Fun! 😄** 

Btw, to help you find it, the main method looks like this:
```Java
public static void main(String[] args) {
         UI UI = new UI();
         String logo = " ____        _        \n"
                 + "|  _ \\ _   _| | _____ \n"
                 + "| | | | | | | |/ / _ \\\n"
                 + "| |_| | |_| |   <  __/\n"
                 + "|____/ \\__,_|_|\\_\\___|\n";
         System.out.println("Hello from\n" + logo);

         UI.start();
     }
```
Duke can do this for you:

- [ ] Manage _Tasks_
- [ ] Manage _Events_
- [ ] Manage _Deadlines_
- [ ] ~~Your University Assignments~~

`System.out.println("Thanks for Reading!")`

Commit:
Level 1 fnished

Commit:
Level 2 finished.

Commit:
Level 3 finished.

Commit:
Level 4 finished.

Commit:
A-TextUiTesting

Commit:
Level 5 finished.

Commit:
Level-6 finished, added more exception handling

Commit:
Save progress before checkout

Commit:
Save progress before checkout

Commit:
Add storage of tasks

Commit:
Fix some merge issues, update directory structure

Commit:
Update directory names

Commit:
Fix merge conflicts

Commit:
Add more OOP

Commit:
Refactor all code under duke package

Commit:
Add JUnit Testing

Commit:
Remove redundant code

Commit:
Add ability to get list of tasks from search using keyword.

Commit:
Add gradle to project

Commit:
Add checkstyle

Commit:
Add fully functional GUI to Duke

Commit:
Rebase checkstyle fixes on master

Commit:
Add attributions

bingsen0806 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+            return input;
+        } else {
+            return input.substring(0, indexOfWhitespace);
+        }
```
Perhaps there is no need for the `else` keyword because the `if` clause already serves as a guard clause. This may help improve readability and code quality.

bingsen0806 commented on src/main/java/duke/exceptions/RequiredInformationMissingException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Exception that occurs when users does not give a description when adding a task.
+ */
+public class RequiredInformationMissingException extends DukeException {
```
Great job splitting your `DukeException` class into more Exception classes in a logical manner!

bingsen0806 commented on src/main/java/duke/CommandExecutor.java:
```diff
@@ -0,0 +1,95 @@
+            taskManager.addTask(task);
+            str.append("Got it. I've added this task:\n");
+            str.append("  " + task + "\n");
+            str.append("Now you have " + taskManager.getNumberOfTasks() + " duke.tasks in the list.");
```
These 3 lines seem to be quite repetitive in this function. Perhaps it is not a very suitable place for this?

Will it be better if you move these 3 lines into your `Task::execute` method and let them be printed out when you execute the tasks?

bingsen0806 commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,50 @@
+     * @throws UnknownCommandException
+     */
+    @Test
+    public void testEventInputParsing() throws RequiredInformationMissingException, UnknownCommandException {
```
Some minor suggestions: perhaps the test method name can be in the form of
`testEventInputParding_whatInput_expectedOutcome` as suggested by the guidelines.

bingsen0806 commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,78 @@
+     * @param isCompleted
+     */
+    public Task(String name, boolean isCompleted) {
+        this.name = name;
```
It's good to see that you only use the `this` keyword when there is clash in name between argument and attribute variables, but not when the `this` keyword is redundant.

Keep it up!

bingsen0806 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+        if (command.equals("list")) {
+            parsedInput.add("list");
+        } else if (command.equals("mark")) {
+            if (input.length() < 6) {
+                throw new RequiredInformationMissingException("missing mark index");
+            }
```
Since this method is a bit too lengthy, I suggest you move all these input checkings into separate small private methods that returns a boolean.

The rationale is that you can name the methods in a way that allows others to understand what it is checking. For example, I do not understand why is the need to check for `input.length() < 6`. Maybe putting it in a method named 'hasInputIndex()` will make more sense. 

Another suggestion is that, perhaps checking length isn't the most suitable way to check for missing index? Maybe you can try string matching to match for numerics only.

bingsen0806 reviewed (commented):
Overall, a great job done! 👍 
I like how you did extensive JavaDocs comments for all your public methods and use reasonable names for your methods.

aceszhenwei commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    @FXML
+    public void initialize() {
+        String introMsg = "Hello, I'm Duke, your trusty doggo assistant!\nHere's what i can do:\n\n"
```
Perhaps this could be renamed to `introMessage` to avoid 'texting-style' spelling.

aceszhenwei commented on src/main/java/duke/TaskManager.java:
```diff
@@ -0,0 +1,118 @@
+     * @param expr the keyword used to search for matching tasks.
+     * @return list of tasks with names matching the expr.
+     */
+    public ArrayList<Task> findListOfMatchingTasks(String expr) {
```
Nice! I have seen that you have used a good habit throughout your code to let your names not just differentiate themselves, but also explain what they do!

aceszhenwei commented on src/main/java/duke/exceptions/UnknownCommandException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exceptions;
+
+/**
+ * Exception thrown when user enters a command that duke.Duke has not been programmed to handle.
+ */
```
Well done on leaving javadocs for your exceptions! This is very helpful especially since these are custom exceptions.

aceszhenwei reviewed (commented):
Really well done! I found your code easy to read and follow. Apart from minor issues that @bingsen0806 may have mentioned, I think you are LGTM!

benmurphyy committed:
Add assert statements to certain points in code

Need to validate certain asseumptions in the code, like in Parser class.

Add assert statements to Parser class to check that assumptions
regarding string length of arguments given to parser are true.

benmurphyy committed:
Fix code quality issues

Various code quality issues were pointed out by reviewers of the PR.

Fixed the code quality issues according to code quality guidelines in the various files of the project.

benmurphyy committed:
Remove log file

benmurphyy committed:
Remove log file

benmurphyy committed:
Merge pull request #1 from benmurphyy/branch-A-Assertions

Add assert statements to certain points in code

benmurphyy committed:
Fix merge conflicts

benmurphyy committed:
Merge pull request #2 from benmurphyy/branch-A-CodeQuality

Fix code quality issues

benmurphyy committed:
Add sort by name function to Duke

Extend functionality of Duke so it can be more useful to the user.

Provide the user the option to sort their items by name so that they
can view them more easily.

benmurphyy committed:
Merge pull request #3 from benmurphyy/branch-A-CI

Add sort by name function to Duke

benmurphyy committed:
Fix chekstyle issues

benmurphyy committed:
Add User guide

benmurphyy committed:
Change readme slightly

benmurphyy committed:
Update gradle build file

benmurphyy committed:
Edit minor typo

benmurphyy committed:
Edit minor typo

benmurphyy committed:
Merge pull request #5 from benmurphyy/branch-A-JavaDoc

Edit minor typo

benmurphyy committed:
Edit minor typo

benmurphyy committed:
Merge pull request #6 from benmurphyy/branch-A-CodingStandard

Edit minor typo
